stdpopsim/genomes.py

```diff
@@ -174,16 +215,154 @@ class Contig:
     genetic_map = attr.ib(default=None, kw_only=True)
     inclusion_mask = attr.ib(default=None, kw_only=True)
     exclusion_mask = attr.ib(default=None, kw_only=True)
+    genomic_element_types = attr.ib(factory=lambda: [], type=list, kw_only=True)
+    mutation_types = attr.ib(factory=lambda: [], type=list, kw_only=True)
+
+    @property
+    def length(self):
+        if self.recombination_map is None:
+            return None
+        return self.recombination_map.sequence_length
+
+    @property
+    def slim_fractions(self):
+        map(lambda g: _check_mut_proportions(g.proportions), self.genomic_element_types)
+        props = np.array([sum(g.proportions) for g in self.genomic_element_types])
+        return props
+
+    @property
+    def genomic_intervals(self):
+        """
+        Returns an (n, 3) numpy array with n intervals (left, right, type)
+        """
+        all_intervals = np.concatenate(
+            [
+                np.column_stack(
+                    (g.intervals[:, :2], np.full((g.intervals.shape[0]), i))
+                )
+                for i, g in enumerate(self.genomic_element_types)
+            ],
+            axis=0,
+        )
+        all_intervals = stdpopsim.utils.build_intervals_array(all_intervals)
+        return all_intervals
+
+    @property
+    def msp_mutation_rate_map(self):
+        """
+        breaks = [0,10,20] means [0,10) and [10,20)
+        """
+        breaks = [0]
+        rates = []
+        for (start, end, t) in self.genomic_intervals:
+            if start not in breaks:
+                breaks.append(start)
+                rates.append(self.mutation_rate)
+            breaks.append(end)
+            rates.append((1 - self.slim_fractions[t]) * self.mutation_rate)
+        if not np.isclose(breaks[-1], int(self.length)):
+            breaks.append(int(self.length))
+            rates.append(self.mutation_rate)
+        assert len(breaks) == len(rates) + 1
+        return (breaks, rates)
+
+    @property
+    def slim_mutation_rate_map(self):
+        """
+        breaks = [10,20] means [0,10] and (10,20]
+        """
+        # making sure the mut types are still concordant with the ge types
+        breaks, rates = self.msp_mutation_rate_map
+        assert breaks[0] == 0
+        slim_breaks = [b - 1 for b in breaks[1:]]
+        slim_rates = [self.mutation_rate - r for r in rates]
+        return (slim_breaks, slim_rates)
+
+    def add_mutation_types(self, mutation_types, proportions, genomic_element_type_id):
+        """
+        Adds mutation types with their respective proportions to the genomic
+        element type provided.
+        """
+        _check_mut_proportions(proportions, mutation_types)
```

Comment:
this check should also include any (potentially) existing proportions and mutation types.

Reply:
done